### ISSUE TITLE
Unexpected PnL when flipping futures position

### ISSUE THREAD

**Issue by dhexus** [quantopian/zipline#783]
Hi. I've a futures algorithm where pnl does not match expected pnl. So I wrote a little test. I use instant_fill=True, no slippage, and no commission. Moreover I provide only three data events of type TRADE, to make it very easy. I use a DataPanelSource that holds the following data for the relevant asset:
``` 
data = pd.DataFrame([
            [1, 1e9, int(DATASOURCE_TYPE.TRADE)],
            [2, 1e9, int(DATASOURCE_TYPE.TRADE)],
            [4, 1e9, int(DATASOURCE_TYPE.TRADE)]],
            columns=['price', 'volume', 'type'], index=index)
```
My handle_data method looks like this:
```
def handle_data(self, data):
        if len(self.portfolio.positions) > 0:
            if self.portfolio.positions[self.asset.sid]["amount"] > 0:
                self.order_target(self.asset, -self.amount)
            else:
                self.order_target(self.asset, 0)
        else:
            self.order_target(self.asset, self.amount)
```
So I go long if there is no position, I go short when I'm long, and I go flat when I'm short. Moreover, I have  amount=1 and multiplier=168. In results object, positions shows exactly the number of contracts that I would expect, i.e. [1,-1,0]. However, pnl is [0, 0, 0]. Ignoring initial margin, I would expect [0, 168, -336].

CC @warren-oneill @jfkirk @StewartDouglas @ssanderson

**Comment by warren-oneill** [comment]
@jfkirk @StewartDouglas have you guys had a chance to look at this? Can you you reproduce our error? Is this behavior already tested for futures in zipline?

**Comment by jfkirk** [comment]
@dhexus @warren-oneill I am attempting to reproduce now and will update when I have reproduced or if I need more information.

Self-assigning.

**Comment by warren-oneill** [comment]
@jfkirk thanks :smile:

**Comment by jfkirk** [comment]
Thanks to Warren's test, I've been able to reproduce the issue. It appears that the issue is caused by the Position's last_sale_price already being updated before PositionTracker.update_last_sale is called for each trade event. This causes the cash flow calculation to return 0s because price == old_price:

```
price = event.price

        if checknull(price):
            return 0

        pos = self.positions[sid]
        old_price = pos.last_sale_price
        pos.last_sale_date = event.dt
        pos.last_sale_price = price

        # Calculate cash adjustment on assets with multipliers
        return ((price - old_price) * self._position_payout_multipliers[sid]
                * pos.amount)
```


I am digging further in to the why.

**Comment by jfkirk** [comment]
I've just confirmed that this anomalous early-update-of-last-sale-price is not happening in other Futures trading algos (for example test_algorithm.py:TestClosePosAlgo.test_close_position_future works a-ok).

Still investigating.

**Comment by warren-oneill** [comment]
@jfkirk but in that test you trade only once. Correct? I think the problem arises when one product is traded multiple times.

**Comment by jfkirk** [comment]
@warren-oneill That's right - it seems the problem is caused by the sequence of events between execute_transaction and update_last_sale. In your algo, the transactions are coming through before the trade events with the new price, so the position is getting updated (both quantity and last sale price) before the new trade event comes through. This means that, when the trade event comes in, the Position thinks that the price hasn't changed, and then we don't receive a cash flow.

I'm figuring out the best way to fix this now.

**Comment by jfkirk** [comment]
This particular issue is fixed with the merge of #803 -- closing issue
